batchflow/models/eager_torch/base.py

```diff
@@ -0,0 +1,1203 @@
+
+        Examples
+        --------
+        .. code-block:: python
+
+            def build_config(self, names=None):
+                config = super().build_config(names)
+                config['head/filters'] = self.num_classes('targets')
+                return config
+        """
+        config = self.full_config
+
+        if config.get('inputs'):
+            inputs_config = config['inputs']
+
+            # Add default aliases
+            if 'targets' not in inputs_config:
+                if 'labels' in inputs_config:
+                    inputs_config['targets'] = inputs_config['labels']
+                elif 'masks' in inputs_config:
+                    inputs_config['targets'] = inputs_config['masks']
+
+            # Fetch default data format for all the parts of the network
+            inputs = config.get('initial_block/inputs')
+            if isinstance(inputs, str):
+                data_format = inputs_config.get(inputs, {}).get('data_format')
+            elif isinstance(inputs, (tuple, list)):
+                data_format = inputs_config.get(inputs[0], {}).get('data_format')
+            else:
+                data_format = 'channels_first'
+            config['common/data_format'] = config.get('common/data_format') or data_format
+
+        config['head/target_shape'] = self.target_shape
+        config['head/classes'] = self.classes
+        return config
+
+
+    def _get_devices(self):
+        devices = self.full_config.get('device')
+        if devices is None:
+            if torch.cuda.is_available():
+                self.device = torch.device('cuda:0')
+            else:
+                self.device = torch.device('cpu')
+        else:
+            devices = self.full_config.get('device')
+            devices = devices if isinstance(devices, list) else [devices]
+            available_devices = ['cuda:{}'.format(i) for i in range(torch.cuda.device_count())] + ['cpu']
+
+            for dev in devices:
+                if isinstance(dev, torch.device):
+                    self.devices.append(dev)
+                elif isinstance(dev, str):
+                    dev_ = dev.lower()
+                    dev_ = dev.replace('gpu', 'cuda')
+
+                    devices = [torch.device(device) for device in available_devices
+                               if re.search(dev_, device.lower()) is not None]
+                    self.devices.extend(devices)
+                else:
+                    raise TypeError('Wrong device type: {}'.format(type(dev)))
+            self.devices = [device for i, device in enumerate(self.devices)
+                            if device not in self.devices[:i]]
+            self.device = self.devices[0]
+
+        torch.backends.cudnn.benchmark = self.full_config.get('benchmark', 'cuda' in self.device.type)
+
+    def _get_placeholder_shapes(self):
+        config = self.full_config
+        batch_size = config.get('placeholder_batch_size', 2)
```

Comment:
move this under `if input_names is not None:`
and specify default values for placeholder_batch_size in docstrings